### PR TITLE
nixos/utils: support JSON secret files in genJqSecretsReplacementSnippet

### PR DESCRIPTION
**Pull request by linyinfeng**
## Description of changes

An alternative to <https://github.com/NixOS/nixpkgs/pull/324955>.

> genJqReplacementSnippet quotes the content of the secret file in the
> output json file, which prevents structured secret, such as a list or
> an object, from being used.

This PR adds a `quote = true|false` option to the `{ _secret = "/path/to/secret"; }` attribute set. `quote = true` threats the content of `/path/to/secret` as string, while `quote = false` threats the content of `/path/to/secret` as a JSON document.

`quote = true` is the default, so this PR maintains backward compatibility.

```nix
# demo.nix

let
  demoNixpkgs = fetchTarball "https://github.com/linyinfeng/nixpkgs/archive/refs/heads/json-secrets.zip";
  pkgs = import demoNixpkgs {};
  inherit (pkgs) lib;
  utils = import "${demoNixpkgs}/nixos/lib/utils.nix" {
    inherit lib pkgs;
    config = null;
  };
  settings = {
    example = [
      {
        irrelevant = "not interesting";
      }
      {
        ignored = "ignored attr";
        relevant = {
          secretString1 = {
            _secret = pkgs.writeText "password1" "secret-quote-1";
            # quote = true is the default
          };
          secretString2 = {
            _secret = pkgs.writeText "password2" "secret-quote-2";
            quote = true;
          };
          secretJson = {
            _secret = pkgs.writeText "password3" (builtins.toJSON {
              p1 = "supersecret";
              p2 = [2 3];
            });
            quote = false;
          };
        };
      }
    ];
  };
  snippet = utils.genJqSecretsReplacementSnippet settings;
in
pkgs.runCommand "test.json" {} ''
  ${snippet "result.json"}
  cp result.json "$out"
''
```

```console
$ nix-build ./demo.nix && cat result
/nix/store/36rgaa21gml1xl0lk3rdm28z3vw1rv84-test.json
{
  "example": [
    {
      "irrelevant": "not interesting"
    },
    {
      "ignored": "ignored attr",
      "relevant": {
        "secretJson": {
          "p1": "supersecret",
          "p2": [
            2,
            3
          ]
        },
        "secretString1": "secret-quote-1",
        "secretString2": "secret-quote-2"
      }
    }
  ]
}
```

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/lib/utils.nix [hunk 1]
```diff
@@ -206,14 +231,69 @@ utils = rec {
...
+       genJqSecretsReplacementSnippet {
+         example = [
+           {
+             irrelevant = "not interesting";
+           }
+           {
+             ignored = "ignored attr";
+             relevant = {
+               secret = {
+                 _secret = "/path/to/secret";
```
Comment (jian-lin) [inline nixos/lib/utils.nix:259]:
```suggestion
                 _secret = "/path/to/secret";
                 quote = false;
```

**Review by jian-lin — COMMENTED** [review]
LGTM

Could you add the context (motivation and other things) into the commit message so that it can be accessed without going to github?

**Comment by jian-lin** [comment]
https://github.com/ofborg test sing-box lemmy engelsystem artalk discourse

filebeat does not have a NixOS test.

netbird's NixOS test does not use secrets.

gitlab's NixOS test is flaky.

File: nixos/lib/utils.nix (hunk 1 above)
Comment (linyinfeng) [reply nixos/lib/utils.nix:259]: Oops, I forgot this, fixed.

**Review by jian-lin — APPROVED** [review]

**Pull request merged**